### PR TITLE
Feature/graphql

### PR DESCRIPTION
**Pull request by Psyroo**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,6 +1,13 @@
 version: "2"
 
 services:
+  hotspot:
```
Comment (thmarinho) [inline docker-compose.yml:4]: should probably remove this service

File: graphql/index.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    connectionParams: {
+      headers: {
+        authorization: TOKEN
+      }
+    }
+  }, ws)
+  return client
+}
+
+const createSubscriptionObservable = (wsurl: string, query: DocumentNode, variables: {[key: string]: string | number}
```
Comment (thmarinho) [inline graphql/index.ts:29]: should define an interface for `variables`

File: graphql/index.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  if (await createBan(msgData.mac, false)) {
+    channel.ack(qMsg)
+  }
+  else {
+    channel.nack(qMsg)
+  }
+}
+
+const main = async (): Promise<void> => {
+  console.log("OUIOUIOUI")
```
Comment (thmarinho) [inline graphql/index.ts:134]: :angry:

File: graphql/index.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  await initRabbitMQ()
+  console.log('RabbitMQ is running')
+  await selfCreate(process.env.BALENA_DEVICE_NAME_AT_INIT, process.env.BALENA_DEVICE_UUID + ".balena-devices.com")
+  console.log(`Router ${id} have been created`)
+  await getBans()
+  console.log("NONNONNON")
```
Comment (thmarinho) [inline graphql/index.ts:140]: :angry:

File: package.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (thmarinho) [inline package.json:1]: what is this ?

**Review by thmarinho — CHANGES_REQUESTED** [review]

**Comment by thmarinho** [comment]
@matfire we should probably sqash this branch during the merge

File: graphql/index.ts (hunk 3 above)
Comment (Psyroo) [reply graphql/index.ts:134]: done

File: graphql/index.ts (hunk 4 above)
Comment (Psyroo) [reply graphql/index.ts:140]: done

File: docker-compose.yml (hunk 1 above)
Comment (Psyroo) [reply docker-compose.yml:4]: done

File: package.json (hunk 5 above)
Comment (Psyroo) [reply package.json:1]: done

File: graphql/index.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    connectionParams: {
+      headers: {
+        authorization: TOKEN
+      }
+    }
+  }, ws)
+  return client
+}
+
+const createSubscriptionObservable = (wsurl: string, query: DocumentNode, variables: variable) => {
```
Comment (thmarinho) [inline graphql/index.ts:29]: should use Variable instead of variable

File: graphql/index.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+interface GraphqlRequestContext {
+  query: string;
+  variables: {
+    [key: string]: {
+      [key: string]: string | number | boolean
+    } | string | number | boolean
+  };
+}
+
+interface variable {
```
Comment (thmarinho) [inline graphql/index.ts:157]: same here

File: graphql/index.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+main()
+
+interface Ban {
+  address: string;
+  banned: boolean;
+}
+
+interface GraphqlRequestContext {
+  query: string;
+  variables: {
```
Comment (thmarinho) [inline graphql/index.ts:150]: use your interface instead of this

**Review by thmarinho — CHANGES_REQUESTED** [review]

File: graphql/index.ts (hunk 6 above)
Comment (Psyroo) [reply graphql/index.ts:29]: done

File: graphql/index.ts (hunk 7 above)
Comment (Psyroo) [reply graphql/index.ts:157]: done

File: graphql/index.ts (hunk 8 above)
Comment (Psyroo) [reply graphql/index.ts:150]: done

Comment (thmarinho) [reply graphql/index.ts:150]: try again

**Review by thmarinho — CHANGES_REQUESTED** [review]

**Review by thmarinho — APPROVED** [review]

**Review by thmarinho — APPROVED** [review]

File: .gitignore [hunk 9]
```diff
@@ -76,40 +83,41 @@ typings/
...
-!.vscode/settings.json
-!.vscode/tasks.json
-!.vscode/launch.json
-!.vscode/extensions.json
-
-### VisualStudioCode Patch ###
-# Ignore all local history of files
-.history
-.vscode/
+# vuepress build output
```
Comment (matfire) [inline .gitignore:105]: useless

File: docker-compose.yml [hunk 10]
```diff
@@ -26,3 +26,19 @@ services:
     network_mode: "host"
     depends_on:
       - rabbitmq
+    labels:
+      io.balena.features.kernel-modules: "1"
+  # firewall:
```
Comment (matfire) [inline docker-compose.yml:31]: remove commentary: not used atm

File: graphql/constant.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+const BASE_URL =
```
Comment (matfire) [inline graphql/constant.ts:1]: try with env variable

File: graphql/constant.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  !process.env.NODE_ENV || process.env.NODE_ENV === "development"
+  ? "http://localhost:3000"
+  : "https://api.server.vincipit.com"
+
+const GRAPHQL_ENDPOINT = `${BASE_URL}/graphql`
+const FIREWALL_ENDPOINT = "http://localhost:5000"
+
+const TOKEN = "Bearer " //enter yout token
```
Comment (matfire) [inline graphql/constant.ts:9]: wtf?

File: graphql/index.ts [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+  address: string;
+  banned: boolean;
+}
+
+interface GraphqlRequestContext {
+  query: string;
+  variables: Variable
+  };
+
+interface Variable {
```
Comment (matfire) [inline graphql/index.ts:153]: change name

File: graphql/index.ts [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+const requestFirewall = async (ban: Ban): Promise<void> => {
+  console.log(`New ${ban.banned ? "ban" : "unban"} for address ${ban.address}`)
+  await axios({
+    method: ban.banned ? 'post' : 'delete',
+    url: `${FIREWALL_ENDPOINT}/rule/${ban.banned ? "ban": "unban"}MAC?address=${ban.address}`
+  })
+}
+/* Utils End */
+
+const selfCreate = async (name: string, url: string): Promise<void> => {
```
Comment (matfire) [inline graphql/index.ts:54]: use async / await instead of promises ?

File: graphql/index.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        name,
+        url
+      }
+    }
+  })
+  .then(res => id = res.data.data.createRouter._id)
+  .catch(() => console.log("Routeur already created"))
+}
+
+const getBans = async (): Promise<void> => {
```
Comment (matfire) [inline graphql/index.ts:68]: same

File: graphql/index.ts [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+      variables: {
+        banCreationData: {
+          address,
+          banned,
+          routerSet : id
+        }
+      }
+    })
+    return true
+  } catch (error) {
```
Comment (matfire) [inline graphql/index.ts:91]: unused variable

check for error code and status; act accordingly

File: graphql/index.ts [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    {routerSet: id}
+  )
+  client.subscribe(eventData => {
+    requestFirewall(eventData.data.banUpdated)
+  })
+}
+
+const initRabbitMQ = async (): Promise<void> => {
+  try {
+    const connection = await amqp.connect({
```
Comment (matfire) [inline graphql/index.ts:109]: use env variables

File: graphql/index.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+    process.exit(1);
+  }
+};
+
+const consumerDhcp = async (qMsg: amqp.ConsumeMessage): Promise<void> => {
+  const msgData = JSON.parse(qMsg.content.toString())
+  if (await createBan(msgData.mac, false)) {
+    channel.ack(qMsg)
+  }
+  else {
```
Comment (matfire) [inline graphql/index.ts:127]: move else after }

File: graphql/package.json [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  "author": "Arthur LE BRETON",
+  "license": "ISC",
+  "dependencies": {
+    "@apollo/client": "^3.3.7",
+    "@types/amqplib": "^0.5.17",
+    "amqplib": "^0.7.0",
+    "apollo-link": "^1.2.14",
+    "apollo-link-ws": "^1.0.20",
+    "axios": "^0.21.1",
+    "express": "^4.17.1",
```
Comment (matfire) [inline graphql/package.json:19]: remove express / express-graphql

File: graphql/package.json [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+    "express-graphql": "^0.12.0",
+    "graphql": "^15.5.0",
+    "graphql-tag": "^2.11.0",
+    "isomorphic-fetch": "^3.0.0",
```
Comment (matfire) [inline graphql/package.json:23]: remove

File: graphql/package.json [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    "loglevel": "^1.7.1",
+    "node-fetch": "^2.6.1",
```
Comment (matfire) [inline graphql/package.json:25]: remove

File: graphql/package.json [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+    "amqplib": "^0.7.0",
+    "apollo-link": "^1.2.14",
+    "apollo-link-ws": "^1.0.20",
+    "axios": "^0.21.1",
+    "express": "^4.17.1",
+    "express-graphql": "^0.12.0",
+    "graphql": "^15.5.0",
+    "graphql-tag": "^2.11.0",
+    "isomorphic-fetch": "^3.0.0",
+    "loglevel": "^1.7.1",
```
Comment (matfire) [inline graphql/package.json:24]: remove

File: graphql/package.json [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+  "main": "index.js",
+  "scripts": {
+    "start": "node ./index.js",
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "Arthur LE BRETON",
+  "license": "ISC",
+  "dependencies": {
+    "@apollo/client": "^3.3.7",
+    "@types/amqplib": "^0.5.17",
```
Comment (matfire) [inline graphql/package.json:14]: move to devDependencies

File: graphql/package.json [hunk 24]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "graphql",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "start": "node ./index.js",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (matfire) [inline graphql/package.json:8]: remove useless script

File: graphql/package.json [hunk 25]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "graphql",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "start": "node ./index.js",
```
Comment (matfire) [inline graphql/package.json:7]: script does not work

File: graphql/src/banQueries.ts [hunk 26]
```diff
@@ -0,0 +1,37 @@
+const gql = require('graphql-tag');
```
Comment (matfire) [inline graphql/src/banQueries.ts:1]: check import with "graphql" package

File: graphql/src/routerQueries.ts [hunk 27]
```diff
@@ -0,0 +1,16 @@
+import gql from 'graphql-tag'
```
Comment (matfire) [inline graphql/src/routerQueries.ts:1]: same

**Review by matfire — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 9 above)
Comment (Psyroo) [reply .gitignore:105]: done

File: docker-compose.yml (hunk 10 above)
Comment (Psyroo) [reply docker-compose.yml:31]: done

File: graphql/index.ts (hunk 13 above)
Comment (Psyroo) [reply graphql/index.ts:153]: done

File: graphql/constant.ts (hunk 12 above)
Comment (Psyroo) [reply graphql/constant.ts:9]: done

File: graphql/constant.ts (hunk 11 above)
Comment (Psyroo) [reply graphql/constant.ts:1]: done

File: graphql/index.ts (hunk 18 above)
Comment (Psyroo) [reply graphql/index.ts:127]: done

File: graphql/package.json (hunk 19 above)
Comment (Psyroo) [reply graphql/package.json:19]: done

File: graphql/package.json (hunk 20 above)
Comment (Psyroo) [reply graphql/package.json:23]: done

File: graphql/package.json (hunk 21 above)
Comment (Psyroo) [reply graphql/package.json:25]: done

File: graphql/package.json (hunk 22 above)
Comment (Psyroo) [reply graphql/package.json:24]: done

File: graphql/package.json (hunk 24 above)
Comment (Psyroo) [reply graphql/package.json:8]: done

File: graphql/src/banQueries.ts (hunk 26 above)
Comment (Psyroo) [reply graphql/src/banQueries.ts:1]: done

File: graphql/src/routerQueries.ts (hunk 27 above)
Comment (Psyroo) [reply graphql/src/routerQueries.ts:1]: done

File: graphql/package.json (hunk 23 above)
Comment (Psyroo) [reply graphql/package.json:14]: done

File: graphql/package.json (hunk 24 above)
Comment (Psyroo) [reply graphql/package.json:8]: done

File: graphql/package.json (hunk 25 above)
Comment (Psyroo) [reply graphql/package.json:7]: done

File: graphql/src/banQueries.ts (hunk 26 above)
Comment (Psyroo) [reply graphql/src/banQueries.ts:1]: done

File: graphql/src/routerQueries.ts (hunk 27 above)
Comment (Psyroo) [reply graphql/src/routerQueries.ts:1]: done

File: graphql/index.ts (hunk 17 above)
Comment (Psyroo) [reply graphql/index.ts:109]: done

File: graphql/index.ts (hunk 14 above)
Comment (Psyroo) [reply graphql/index.ts:54]: done

File: graphql/index.ts (hunk 15 above)
Comment (Psyroo) [reply graphql/index.ts:68]: done

**Review by matfire — APPROVED** [review]

File: graphql/constant.ts [hunk 28]
```diff
@@ -0,0 +1,16 @@
+const BASE_URL = "URL_SERVER"
```
Comment (matfire) [inline graphql/constant.ts:1]: ?

File: graphql/package.json [hunk 29]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "graphql",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "author": "Arthur LE BRETON",
```
Comment (matfire) [inline graphql/package.json:6]: start script

**Review by matfire — CHANGES_REQUESTED** [review]

File: graphql/package.json (hunk 29 above)
Comment (thmarinho) [inline graphql/package.json:6]: Build script as well

File: graphql/package.json [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+  "license": "ISC",
+  "dependencies": {
+    "@apollo/client": "^3.3.7",
+    "amqplib": "^0.7.0",
+    "apollo-link": "^1.2.14",
+    "apollo-link-ws": "^1.0.20",
+    "axios": "^0.21.1",
+    "graphql": "^15.5.0",
+    "graphql-tag": "^2.11.0",
```
Comment (thmarinho) [inline graphql/package.json:15]: remove that, import `gql` from the `graphql` package

**Review by thmarinho — CHANGES_REQUESTED** [review]

File: graphql/package.json (hunk 29 above)
Comment (Psyroo) [reply graphql/package.json:6]: done

Comment (Psyroo) [reply graphql/package.json:6]: done

File: graphql/constant.ts (hunk 28 above)
Comment (Psyroo) [reply graphql/constant.ts:1]: done

File: graphql/package.json (hunk 30 above)
Comment (thmarinho) [reply graphql/package.json:15]: resolve, can't remove this package

File: graphql/index.ts [hunk 31]
```diff
@@ -0,0 +1,174 @@
...
+    connectionParams: {
+      headers: {
+        authorization: TOKEN
+      }
+    }
+  }, ws)
+  return client
+}
+
+const createSubscriptionObservable = (wsurl: string, query: DocumentNode, variables: queryContent) => {
```
Comment (thmarinho) [inline graphql/index.ts:29]: Type should begin with an Uppercase character

File: graphql/index.ts [hunk 32]
```diff
@@ -0,0 +1,174 @@
...
+      query: print(CREATE_ROUTER),
+      variables: {
+        createRouterData: {
+          name,
+          url
+        }
+      }
+    })
+    id = res.data.data.createRouter._id
+  } catch(error) {
```
Comment (thmarinho) [inline graphql/index.ts:66]: unused variable

File: graphql/index.ts [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+  try {
+    const connection = await amqp.connect({
+      hostname: process.env.AMQP_HOSTNAME,
+      username: process.env.AMQP_USERNAME,
+      password: process.env.AMQP_PASSWORD,
+    });
+    channel = await connection.createChannel();
+    await channel.assertQueue("dhcp");
+    channel.consume("dhcp", consumerDhcp)
+  } catch (error) {
```
Comment (thmarinho) [inline graphql/index.ts:135]: unused variable

File: graphql/index.ts [hunk 34]
```diff
@@ -0,0 +1,174 @@
...
+  address: string;
+  banned: boolean;
+};
+
+interface GraphqlRequestContext {
+  query: string;
+  variables: queryContent
+  };
+
+interface queryContent {
```
Comment (thmarinho) [inline graphql/index.ts:170]: uppercase too

File: graphql/package.json [hunk 35]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "graphql",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (thmarinho) [inline graphql/package.json:5]: sure ?

File: graphql/package.json [hunk 36]
```diff
@@ -0,0 +1,26 @@
...
+  "scripts": {
+    "build": "tsc",
+    "start": "node dist/index.js"
+  },
+  "author": "Arthur LE BRETON",
+  "license": "ISC",
+  "dependencies": {
+    "@apollo/client": "^3.3.7",
```
Comment (thmarinho) [inline graphql/package.json:13]: are you using apollo ?

File: graphql/package.json [hunk 37]
```diff
@@ -0,0 +1,26 @@
...
+    "amqplib": "^0.7.0",
+    "apollo-link": "^1.2.14",
```
Comment (thmarinho) [inline graphql/package.json:15]: same

File: graphql/package.json [hunk 38]
```diff
@@ -0,0 +1,26 @@
...
+    "apollo-link-ws": "^1.0.20",
```
Comment (thmarinho) [inline graphql/package.json:16]: same

**Review by thmarinho — CHANGES_REQUESTED** [review]

File: graphql/index.ts (hunk 31 above)
Comment (Psyroo) [reply graphql/index.ts:29]: done

File: graphql/index.ts (hunk 32 above)
Comment (Psyroo) [reply graphql/index.ts:66]: done

File: graphql/index.ts (hunk 33 above)
Comment (Psyroo) [reply graphql/index.ts:135]: done

File: graphql/package.json (hunk 37 above)
Comment (thmarinho) [reply graphql/package.json:15]: used, resolve

File: graphql/package.json (hunk 38 above)
Comment (thmarinho) [reply graphql/package.json:16]: used as well

File: graphql/index.ts (hunk 34 above)
Comment (Psyroo) [reply graphql/index.ts:170]: done

File: graphql/package.json (hunk 35 above)
Comment (Psyroo) [reply graphql/package.json:5]: done

File: graphql/package.json (hunk 36 above)
Comment (Psyroo) [reply graphql/package.json:13]: done

**Review by thmarinho — APPROVED** [review]

**Review by matfire — APPROVED** [review]

**Pull request merged**